### PR TITLE
Deprecate layer.status

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR deprecates layer.status, replacing it with `layer.get_status(position, world=False)` which takes a position in world coordinates or data coordinates depending on the `world` flag. It also makes the an optional `position` argument for `layer.get_value` required and adds a `world` flag, `layer.get_value(position, world=False)`.

The reason for doing this is that it takes us one step closed to being able to remove `layer.value`, `layer.coordinates`, `layer.displayed_coordinates`, `layer.position` all of which require knowledge of the current cursor position and should not live on the layer, as they will prevent us from supporting multiple canvases.

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Comment by sofroniewn** [comment]
@jni this is ready for review. it would be nice to get this in as it will make #2145 that @VolkerH has requested much easier to implement i think!

File: examples/mouse_drag_callback.py [hunk 1]
```diff
@@ -17,10 +17,6 @@ def profile_lines(image, shape_layer):
                 image, line[0], line[1], mode='reflect'
             ).mean()
         )
-    msg = ('profile means: ['
-            + ', '.join([f'{d:.2f}' for d in profile_data])
-            + ']')
-    shape_layer.status = msg
```
Comment (jni) [inline examples/mouse_drag_callback.py]: do we want to print here or something? Printing is clunky, but this example does ~nothing now without it... I know eventually we'll move to a general notification system — maybe include a comment above the print to that effect?

File: napari/components/viewer_model.py [hunk 2]
```diff
@@ -458,7 +456,7 @@ def _on_cursor_position_change(self, event):
 
         # Update status and help bar based on active layer
         if self.active_layer is not None:
-            self.status = self.active_layer.status
+            self.status = self.active_layer.get_status(self.cursor.position)
```
Comment (jni) [inline napari/components/viewer_model.py]: Ah, nice! 😊 This is very much the key line in this PR! LOL It makes it clear where we're going with this.

One suggestion to consider, not 100% about it:

```python
layer.get_status(self.cursor.position)  # value from world coordinates
layer.get_status(world_pos=self.cursor.position)  # same as above
layer.get_status(data_pos=data_coord)  # value from data coordinates
```

Thoughts?  Maybe the bool flag is better, just throwing this out there.

Comment (jni) [reply napari/components/viewer_model.py]: Or maybe an enum with `world`, `canvas`, and `data` as the possible values?

File: napari/layers/labels/labels.py [hunk 3]
```diff
@@ -858,18 +856,35 @@ def paint(self, coord, new_label, refresh=True):
...
+            dimensions will be used.
+
+        Returns
+        -------
+        msg : string
+            String containing a message that can be used as a status update.
+        """
+        data_position = self._world_to_data(position)
+        value = self.get_value(data_position)
+        msg = genetate_layer_status(self.name, data_position, value)
```
Comment (jni) [inline napari/layers/labels/labels.py]: heh, missed this other places — genetate -> generate

**Review by jni — APPROVED** [review]
@sofroniewn this needs a typo fix, otherwise good from me. My comment about possible APIs going forward does not need to be resolved in this PR.

File: napari/components/viewer_model.py (hunk 2 above)
Comment (sofroniewn) [reply napari/components/viewer_model.py]: I'm open, does @tlambert03 have a recommendation?

File: napari/layers/labels/labels.py (hunk 3 above)
Comment (sofroniewn) [reply napari/layers/labels/labels.py]: lol, will fix

File: examples/mouse_drag_callback.py (hunk 1 above)
Comment (sofroniewn) [reply examples/mouse_drag_callback.py]: fair point, this ability is now lost. Curious how @tlambert03 things about this in the context of the general notification system. I hadn't really imagined that being used for status things like this (I'd seen it more for warnings/ errors) but it's an interesting idea. Input on the rest of the PR is fully welcome too @tlambert03

Comment (tlambert03) [reply examples/mouse_drag_callback.py]: This is ultimately what sets the status at the bottom of the main window right?  I agree, I hadn't imaged the general notifications for that either.  my first thought is that should be a slot on the qt_main_window that other things can connect to

File: napari/components/viewer_model.py (hunk 2 above)
Comment (tlambert03) [reply napari/components/viewer_model.py]: I like the idea in general.  I think I personal like the enum/string

**Review by tlambert03 — APPROVED** [review]
don't have time for a thorough review now, but I very much like what I'm seeing in `base.py`.  👍

File: examples/mouse_drag_callback.py (hunk 1 above)
Comment (sofroniewn) [reply examples/mouse_drag_callback.py]: yes, correct it is for that status bar. You can still get the status to update with just `viewer.status = 'something new'`, but this PR wouldn't let you also do `layer.status='something new'`

Comment (tlambert03) [reply examples/mouse_drag_callback.py]: yeah, I get (and like) that.  the layer shouldn't have a persistent status.

Comment (sofroniewn) [reply examples/mouse_drag_callback.py]: ok, i've update to have a print, we do this in other examples like `custom_mouse_functions.py`

File: napari/components/viewer_model.py (hunk 2 above)
Comment (sofroniewn) [reply napari/components/viewer_model.py]: So looking at this more the layer can never accept `canvas` as it doesn't have the canvas2world transform info on it, so it's just `world` or `data`. I've made status take a `world=bool` now too, and made it kwarg only, same as for `get_value`. The defaults for both are `world=False`, which has pros/ cons, but I think is fine

**Comment by sofroniewn** [comment]
Thanks for the review @jni and comments @tlambert03 - I've made some minor updates and fixed the typo. I made the `world=False` parameter kwarg only for both `get_status` and `get_value` as we never have a third `canvas` option. If people don't like this then let me know! Otherwise I'll merge tomorrow :-)

**Comment by jni** [comment]
> as we never have a third canvas option

But are you sure we will *never* have a third option? =) I don't foresee it either, but I'm not very experienced with UI design — maybe some people will want to use canvas/screen coordinates for something?

At any rate, keyword-only arguments are easy to deprecate so we can definitely just cross that bridge when we get to it.

**Comment by sofroniewn** [comment]
> But are you sure we will never have a third option? =) I don't foresee it either, but I'm not very experienced with UI design — maybe some people will want to use canvas/screen coordinates for something?

I could imagine a `viewer.get_status(layer, position, 'canvas')`, but it's not something that I think we'd ever allow for the `get_status` method on the layer itself because of the `canvas2world` transform information isn't available there.

> At any rate, keyword-only arguments are easy to deprecate so we can definitely just cross that bridge when we get to it.

Sounds good!

**Pull request merged**